README.md
# Typhoon

The minimalist blog engine for Node.js

# How it works

- Typhoon is built right on top of **[Express](http://expressjs.com)** making it easy to extend
- Content is entirely managed through **git**
- Articles are written in _.txt_ with embedded metadata (in **yaml** format)
- Articles are to be written in **Markdown** format and are compiled with  [node-markdown](//github.com/andris9/node-markdown)
- Templating is done using **Jade**
- Comments are handled by [disqus](http://disqus.com/)
- Typhoon's RSS feed can be used with [FeedBurner](http://feedburner.com/)
- Individual articles can be accessed through urls such as _/2011/03/28/hello-world_
- Archives can be accessed by year, month or day, with the same format as above

# Obtaining Typhoon

The easiest way to obtain Typhoon is through [npm](http://npmjs.org/):

    npm install typhoon

# Setting up your blog

If you are anxious to get started simply clone the repository [cjoudrey/blog.ht4.ca](https://github.com/cjoudrey/blog.ht4.ca) which comes with everything to get you started.

    git clone git@github.com:cjoudrey/blog.ht4.ca.git

Otherwise you can set up your blog manually as follow.

The first thing you will need to get started is the following directory structure:

    blog/
    |-- articles/          # Contains your articles
    |-- public/            # Contains your assets (css, images, favicon)
    |-- views/             # Contains your jade views
      |-- article.jade     # Used to render an article
      |-- error.jade       # Used when there is a http error (404, 500) in production
      |-- layout.jade      # Frame that is used to render each content page
      |-- list.jade        # Used to render a list of articles
      |-- feed.jade        # Used to render the RSS feed (optional)
    |- configs.js          # Typhoon configurations
    |- index.js            # Node entry-point

A full documentation of available locals for each view is available [here](//github.com/cjoudrey/typhoon/blob/master/docs/views.md).

# Node entry-point - index.js

The node entry-point is the script you will be executing with node.

    var typhoon = require('typhoon');
    typhoon.app(__dirname, require('./configs');

# Typhoon configurations - configs.js

The configurations file is a module that exports an object with the following format:

    module.exports = {
      'env': 'production',                          # this can either be 'production' or 'dev'
                                                    # and is used to hide/show errors.

      'title': 'my blog',                           # blog title
      'description': 'just another blog',           # blog description
      'favicon': __dirname + '/public/favicon.ico', # path to the blog's favicon
      'staticDir': __dirname + '/public',           # path to the blog's assets folder
      'viewsDir': __dirname + '/views',             # path to the views
      'articlesDir': __dirname + '/articles',       # path to the articles
      'host': '127.0.0.1',                          # host to listen on
      'port': 8080,                                 # port to listen on
      'baseUrl': 'http://127.0.0.1:8080',           # base url
      'encoding': 'utf8',                           # encoding of the articles
      'perPage': 5,                                 # articles per page

      # Optional configurations
      'articlesExt': '.txt',                        # extension of article files
      'viewsEngine': 'jade',                        # views engine
      'rss': true,                                  # enable the rss feed (requires feed view)

      # Specific to the views used by blog.ht4.ca
      'googleAnalytics': 'UX-XXXXX-X',              # google analytics tracking code
      'disqus': 'myblog',                           # disqus site id
      'feedburner': 'myblog'                        # feedburner site id
    }

# Article structure

Each article placed in the `articles/` folder must have a filename with the following format: `YYYY-MM-DD-slug.txt`.

The filename is used to date the article and to build a link to the article.

The content of the file is formed by a metadata section and the article's body separated by an empty line `/\n\n/`.

The only required metadata is `title`. Additional meta tags can be added and will be accessible in the views.

Example article `2011-04-03-lorem-ipsum.txt`:

    title: Lorem ipsum
    author: chris

    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    Maecenas justo neque, dictum eget accumsan non, luctus ac lacus.
    Phasellus ac erat metus, et sagittis dolor.

This article can now be accessed through `/2011/04/03/lorem-ipsum`.

One can specify a summary for the article by placing the `<!-- more -->` delimiter in the article's body:

    title: Lorem ipsum
    author: chris

    Lorem ipsum dolor sit amet, consectetur adipiscing elit.<!-- more -->
    Maecenas justo neque, dictum eget accumsan non, luctus ac lacus.
    Phasellus ac erat metus, et sagittis dolor.

# Article static files

In some situations one may need static files for an article. These are are to be placed within a folder named after the article.

For instance, the static files folder for the article `2011-04-03-lorem-ipsum.txt` is `2011-04-03-lorem-ipsum`.

This folder is to be placed in the `articles/` folder.

All files placed inside the `articles/2011-04-03-lorem-ipsum/` folder can then be accessed through `/2011/04/03/lorem-ipsum/filename`.

# Documentation for Views

  Local variables are documented [here](//github.com/cjoudrey/typhoon/blob/master/docs/views.md).

# View helpers

  Documentation available [here](//github.com/cjoudrey/typhoon/blob/master/docs/helpers.md).

# Blogs using Typhoon

- [http://blog.ht4.ca](http://blog.ht4.ca) ([Source](//github.com/cjoudrey/blog.ht4.ca))
- [http://blog.clipboard.com/](http://blog.clipboard.com/)

# Mentions

Typhoon was inspired by [Toto](//github.com/cloudhead/toto) (Ruby) and [Wheat](//github.com/creationix/wheat) (Node.js). Go check them out!

Copyright (c) 2011 Christian Joudrey. See LICENSE for details.

Node.js is an official trademark of Joyent. This module is not formally related to or endorsed by the official Joyent Node.js open source or commercial project.


CHANGELOG.md
0.4.1 / 2011-11-08
==================

  * Enabled support for node 0.6.0
  * Updated Express to 2.5.0

0.4.0 / 2011-04-29
==================

  * Added static file folders for articles - closes #8
  * Added tests - closes #3
  * Updated Express to 2.3.2

0.3.3 / 2011-04-17
==================

  * Release to fix NPM package

0.3.2 / 2011-04-17
==================

  * Fixed an issue with Article.permalink relative argument.

0.3.1 / 2011-04-17
==================

  * Added Jade to dependencies
  * Fixed an issue with paging not working properly

0.3.0 / 2011-04-17
==================

  * Complete recode to [Express](http://expressjs.com)
  * Default template engine now [Jade](http://jade-lang.com)

0.2.2 / 2011-04-09
==================

  * Fixed an issue with an uninitialized variable

0.2.1 / 2011-04-09
==================

  * Fixed a typo in `view.coffee`

0.2.0 / 2011-04-09
==================

  * Major rehaul of the API
  * Modified the `summary` helper. Can now specify `delimiter` and `trimmer`
  * Removed `feed.haml` and added optional configuration `configs.rss = true`
  * Helpers are now in their own file `helpers.coffee`
  * Removed `patches.coffee` now `utils.coffee`
  * Added optional configurations: `articlesExt`, `templatesExt`

0.1.2 / 2011-04-05
==================

  * Added date validation for listings - closes #7
  * Added template local variable `title` in archives - closes #2
  * Added gravatar helper `gravatar(email, size = 50)`
  * Template helpers are now exposed in `Typhoon.Helpers`

0.1.1 / 2011-04-03
==================

  * Removed src from published package

0.1.0 / 2011-04-03
==================

  * Initial release


src/typhoon/app.js
var express = require('express');
var Article = require('./models/article').Article;
var helpers = require('./helpers');

module.exports.app = function(root, configs, listen) {
  var app = express.createServer();

  configs = module.exports.setDefaultConfigs(root, configs);

  app.helpers(helpers);
  app.dynamicHelpers({
    configs: function() {
      return configs;
    }
  });

  app.configure(function() {
    for (key in configs) {
      app.set('typhoon ' + key, configs[key]);
    }

    Article.includePath = app.set('typhoon articlesDir');
    Article.extension = app.set('typhoon articlesExt');
    Article.baseUrl = app.set('typhoon baseUrl');
    Article.encoding = app.set('typhoon encoding');

    app.set('views', app.set('typhoon viewsDir'));
    app.set('view engine', app.set('typhoon viewsEngine'));
    app.use(express.bodyParser());
    app.use(express.methodOverride());

    if (app.set('typhoon favicon')) { 
      app.use(express.favicon(app.set('typhoon favicon')));
    }

    if (app.set('typhoon staticDir')) {
      app.use(express.static(app.set('typhoon staticDir')));
    }

    app.use(app.router);
  });

  app.configure('development', function() {
    app.use(express.logger());
    app.use(express.profiler());
    app.use(express.responseTime());
    app.use(express.errorHandler({ dumpExceptions: true, showStack: true })); 
  });

  app.configure('production', function() {
    app.enable('view cache');
  });

  require('./routes/article').setup(app);

  app.error(function(err, req, res, next) {
    var locals = {};
    
    locals.errorCode = 500;
    locals.layout = false;
    
    if (err instanceof Error && Number(err.message) === 404) {
      locals.errorCode = 404;
    }
    
    res.render('error', locals);
  });

  if (listen !== false) {
    app.listen(app.set('typhoon port'), app.set('typhoon host'));
  }
  
  return app;
};

module.exports.setDefaultConfigs = function(root, configs) {
  //  Setup default configs
  configs = configs || {};

  // Default blog title is `'untitled blog'`
  configs.title = configs.title || 'untitled blog';

  // Default blog description is `''`
  configs.description = configs.description || '';

  // Default articles directory is `root + '/articles'`
  configs.articlesDir = configs.articlesDir || root + '/articles';

  // Default listen host is `'127.0.0.1'`
  configs.host = configs.host || '127.0.0.1';

  // Default listen port is `8080`
  configs.port = configs.port || 8080;

  // Default base url is `'http://' + configs.host + ':' + configs.port`
  configs.baseUrl = configs.baseUrl || 'http://' + configs.host + (parseInt(configs.port) != 80 ? ':' + configs.port : '');

  // Default articles and templates encoding set to `'utf8'`
  configs.encoding = configs.encoding || 'utf8';

  // Default paging is `10`
  configs.perPage = Number(configs.perPage) || 10;

  // Default article extension is `'.txt'`
  configs.articlesExt = configs.articlesExt || '.txt'

  // Default template extension is `'jade'`
  configs.viewsEngine = configs.viewsEngine || 'jade';
  
  // Default views directory extension is `root + '/views'`
  configs.viewsDir = configs.viewsDir || root + '/views';

  // (optional) Favicon path can be set with:
  // `configs.favicon = root + '/public/favicon.ico'`
  configs.favicon = configs.favicon || false;

  // (optional) Static files directory can be set with:
  // `configs.staticDir = root + '/public/favicon.ico'`
  configs.staticDir = configs.staticDir || false;

  // (optional) Enable RSS feed (requires feed template)
  configs.rss = configs.rss || false;
  
  return configs;
};
